dial/pipeline.go

```diff
@@ -0,0 +1,313 @@
+package dial
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"errors"
+
+	"github.com/libp2p/go-libp2p-net"
+	"github.com/libp2p/go-libp2p-peer"
+	"github.com/libp2p/go-libp2p-transport"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	requestInflight = iota
+	requestFinished
+
+	jobInflight = iota
+	jobFinished
+)
+
+var ErrDialRequestInvalidStatus = errors.New("invalid dial request status")
+var ErrDialJobInvalidStatus = errors.New("invalid dial job status")
+
+// Request represents a request from consumer code to dial a peer.
+// TODO: some of these fields need to be exported to enable dial components to be implemented outside this package.
+type Request struct {
+	net    inet.Network
+	status int32
+
+	// ctx is the parent context covering the entire request.
+	// dial job preparers may stem child contexts from this ctx.
+	ctx context.Context
+
+	// The request starts with the peer ID only.
+	id peer.ID
+
+	// Addresses to be populated by RequestPreparers.
+	addrs []ma.Multiaddr
```

Comment:
I did grapple with the idea of making this a `chan` or a slice, and settled on a slice because I thought the `Planner` needs to know all addresses before it can make sound planning decisions. However, that comes at a time cost: you can't start dialing suboptimal addresses while you fetch better ones in parallel.

I had originally thought about an `AddressResolver` component, with several implementations:

* DefaultAddressResolver => fetches from the peerstore.
* DHTAddressResolver => queries the DHT if there are no hits in the peerstore, or if they are stale.

I'll move back in that direction, as it provides a better encapsulation for the `chan` variant.